### ISSUE TITLE
Add callSuper for RequiredArgsConstructor

### ISSUE THREAD

**Issue by Paramood** [projectlombok/lombok#2916]
It would be nice to be able to write:
```java
@RequiredArgsConstructor
class BaseClass {

    private final String baseClassField;

}

@RequiredArgsConstructor(callSuper = true) // or let callSuper work implicitly
class ExtendedClass extends BaseClass {

    private final Integer extendedClassField;

}
```
instead of
```java
@RequiredArgsConstructor
class BaseClass {

    private final String baseClassField;

}

@RequiredArgsConstructor
class ExtendedClass extends BaseClass {

    private final Integer extendedClassField;

    ExtendedClass(String baseClassField, Integer extendedClassField) {
        super(baseClassField);
        this.extendedClassField = extendedClassField;
    }

}
```

**Comment by Vyacheslav-Lapin** [comment]
> // or let callSuper work implicitly

Maybe (for backward-compatibility) with some setting in _lombok.config_ like we already have with `@EqualsAndHashCode` and `@ToString`?

I.e., something like that:
`lombok.anyConstructor.callsuper = true`

**Comment by janrieke** [comment]
Duplicate of #2646 and many others.
TLDR: Very complicated (maybe impossible) to implement and maintain.

**Comment by claudenobs** [comment]
This is definitively not impossible. Its just not easy. The org.immutables library has been doing this reliably for almost a decade. We have been leveraging lombok & org.immutables for a long time. They work.
So it would be nice if we could also have mutable objects that can generate constructors that call super.

Obviously it's non trivial, a simple solution would be to only allow callSuper if the superclass has only one constructor or it has one by the same annotation (@RequriedArgsConstructor would call the @ReqiredArgsConstructor of super)

**Comment by janrieke** [comment]
org.immutables has a fundamentally different approach in comparison to lombok. Immutables is a standard annotation processor that generates _new_ classes. In contrast, lombok modifies _existing_  classes.
It's obviously a lot easier for Immutables to generate a super-calling constructor if you have the code of the superclass at hand. Lombok does _not_  have it (at least not in all cases).

**Comment by Gouvernathor** [comment]
@janrieke then why not enable it for when the super constructor is a lombok one ? That's an acceptable limitation in SuperBuilder, it would be so here. And as @claudenobs said :
> a simple solution would be to only allow callSuper if the superclass [...] has one by the same annotation (@RequiredArgsConstructor would call the @RequiredArgsConstructor of super)

**Comment by janrieke** [comment]
It only works for `@SuperBuilder` because all its constructors are very similar: the builder is the only parameter. The only thing you can do wrong is forgetting to put `@SuperBuilder` on the superclass, and it will always result in a compilation error. No need to look up anything in the superclass.

This obviously does not work for constructors with parameters: You have to look up the superclass to find out which parameters exist on the constructor. Even if Lombok had already processed the superclass (which may not be the case), it's still difficult to identify the correct one. Remember that javac has not yet performed type resolution when Lombok runs.

**Comment by Gouvernathor** [comment]
Ok, that makes sense. I didn't realize SuperBuilder worked that way, or what kind of restrictions javac could impose.

I'm not sure what type resolution entails exactly, but what I could picture is a sort of caching behavior, where if you encounter a RequiresArgsConstructor on a class, you store (say) the class object in a map about RACs and where the value is the list of parameters, or the constructor itself somehow. If you encounter a "super"-RAC like we propose on a class, then you store in another map the class and the superclass of the decorated type (as key and value in either order).
At each of the two storage actions, you check the other container(s) for the class you're decorating : if your superclass is not object and you are processing a "super"-RAC, you check if the superclass you need the arguments for is already stored in the first map. In any case, you check if there is in a second map a subclass of yourself that's waiting for your information to finalize its own "super"-RAC. When a match is found, you call the superRAC finalizer by passing it the subclass, the superclass, and the arguments list. (And you pop the second map, I guess.)
That would kill two birds with one stone : the first map tells you which of the constructors to use, and the caching mechanism allows you to process the two classes in any order.
And at the end of the compilation process, you find a way so that if the second map is not empty (if a superRAC is used on a subclass where the superclass has no RAC), an error is thrown.

I'm being naïve and talking about something I know almost nothing about, so I fully expect you to tell me it couldn't work ! ^^
